Src/RE2.Managed/Regex2.cs

```diff
@@ -207,9 +220,30 @@ public static unsafe bool Matches(string pattern, string text, out List<Dictiona
                 // Get or Cache the Regex on the native side and retrieve an index to it
                 int expressionIndex = BuildRegex(cache, pattern, RegexOptions.None, maxMemoryInBytes);
 
+                // Cache the expression8, indexMap and the buffer for a given file text. If the text is the same,
+                // load from cache to avoid expensive recomputation which was harming performance.
+                String8 expression8 = String8.Empty;
                 byte[] buffer = null;
-                var expression8 = String8.Convert(text, ref buffer);
-                int[] indexMap = GetMapOfUtf8ToUtf16ByteIndices(buffer);
+                int[] indexMap = null;
+                bool saveToDictionary = false;
+                bool updateIndexMap = false;
+
+                foreach (KeyValuePair<string, Tuple<String8, byte[], int[]>> entry in textToRE2DataMap)
+                {
+                    if (entry.Key.Length == text.Length && entry.Key == text)
+                    {
+                        expression8 = entry.Value.Item1;
+                        buffer = entry.Value.Item2;
+                        indexMap = entry.Value.Item3;
+                        break;
+                    }
+                }
```

Comment:
You could replace your foreach loop with a dict.TryGetValue.
If it exists, just populate expression8, buffer, and indexMap.
If it does not exist, then, you create expression8.

Just to be aware, I'm not saying that this will be faster than what you currently have.

Reply:
Right. Actually, I suggest that you convert this code to use a HybridDictionary, which will use a list internally until the dictionary count exceeds 10 (which shouldn't happen in our utilization!).

Reply:
Looks like maybe there isn't a generic hybrid dictionary?

next suggestion, convert this data structure to an actual list of tuples (and add the scan input string as a new tuple record member).